### ISSUE TITLE
'Rate limit' of new sign-ups from the same IP address

### ISSUE THREAD

**Issue by rodti** [tootsuite/mastodon#16043]
I run two instances with open registration. I regularly see cases of (presumably automated) multiple account sign-ups from the same IP address, which with very few exceptions are spam.

My suggestion is to implement a check against IP address when a new user attempts to register. If that IP address has been used to sign up for another account within a reasonable time frame (1 hour?) then the request is denied, or placed in the registration approval queue.

### Motivation

This proactive check would prevent large numbers of spambots entering the fediverse before moderators spot these accounts or receive reports relating to them, and reduce the effort required to moderate/administer an instance as currently these accounts have to be individually suspended which is an onerous process.

I'd previously raised [this](https://github.com/tootsuite/mastodon/issues/14671) suggestion, however I believe this one is a more elegant solution.

![Screenshot 2021-04-15 at 10 01 45](https://user-images.githubusercontent.com/93992/114844544-cf354580-9dd2-11eb-847c-fcd34e6408bc.png)

**Comment by Gargron** [comment]
Create an IP rule with severity “limit sign-ups” to have them be placed in the moderation queue

**Comment by rodti** [comment]
> Create an IP rule with severity “limit sign-ups” to have them be placed in the moderation queue

That's fine in some scenarios. Here the damage is already done by the time I spot them. 30 user accounts from one IP address, then more from another, and another, and so on. By the time I even realise this has happened there's a huge mess of accounts O need to suspend one by one.

The only solutions I can see are to proactively prevent multiple rapid sign ups from a single IP, or to set my instance to approval-only or closed registration, which seems a shame.

**Comment by ThisIsMissEm** [comment]
We do interestingly have different rate limits for Signup Attempts via the Web and Signup Attempts via the API. For the API it's 5 attempts in 10 minutes, for web it's 25 in 5 minutes.

Perhaps these should be aligned? I'm not sure why they're so different?

**Comment by ThisIsMissEm** [comment]
This is probably best solved by the new "Emergency Mode" feature that's proposed in https://github.com/mastodon/mastodon/pull/25032 where on open registration instances, signups will automatically become restricted if a potential abuse is detected.

**Comment by velw** [comment]
The problem with it being triggered only once the instance is in some kind of emergency mode is that you need to have an emergency for this abuse to be contained.

If I'm understanding correctly, it seems like #31532 was closed without being implemented?

A threshold of 25 sign-up attempts from the same IP address within five minutes seems very high. This seems like a straightforward issue to address, and one that should be addressed.
